File: leetcode/python/0002_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    if l1 == None: return l2
    if l2 == None: return l1

    carry = 0
    res = ListNode()
    n = res
    while l1 or l2:
      val1, val2 = 0, 0

      if not l1: val1 = 0
      else: val1 = l1.val
      if not l2: val2 = 0
      else: val2 = l2.val

      if carry == 0:
        if val1 + val2 < 10: n.val = (val1 + val2) % 10
        else:
          n.val = (val1 + val2) % 10
          carry = 1
      elif carry == 1:
        if val1 + val2 + carry < 10:
          n.val = (val1 + val2 + carry) % 10
          carry = 0
        else: n.val = (val1 + val2 + carry) % 10

      if l1: l1 = l1.next
      if l2: l2 = l2.next
      if l1 or l2: 
        n.next = ListNode()
        n = n.next
    
    if carry == 1: n.next = ListNode(carry)

    return res

File: leetcode/cpp/0002_add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int size1 = 0, size2 = 0;

        ListNode* tmp1Last, * tmp2Last;

        for (tmp1Last = l1; ; tmp1Last = tmp1Last->next) {
            size1++;
            
            if (tmp1Last->next == nullptr) {
                break;
            }
        }

        for (tmp2Last = l2; ; tmp2Last = tmp2Last->next) {
            size2++;
            
            if (tmp2Last->next == nullptr) {
                break;
            }
        }

        if (size1 > size2) {
            for (int i = 0; i < size1 - size2; i++) {
                tmp2Last->next = new ListNode(0, nullptr);
                tmp2Last = tmp2Last->next;
            }
        }
        else if (size2 > size1) {
            for (int i = 0; i < size2 - size1; i++) {
                tmp1Last->next = new ListNode(0, nullptr);
                tmp1Last = tmp1Last->next;
            }
        }

        ListNode* ret = new ListNode(0);

        ListNode& head = *ret;

        for (ListNode* tmp1 = l1, *tmp2 = l2; tmp1 != nullptr && tmp2 != nullptr; tmp1 = tmp1->next, tmp2 = tmp2->next) {
            if (tmp1->val + tmp2->val + ret->val <= 9) {
                ret->val += tmp1->val + tmp2->val;

                if (tmp1->next == nullptr || tmp2->next == nullptr) {
                    break;
                }

                ret->next = new ListNode(0, nullptr);
                ret = ret->next;
            }
            else if (tmp1->val + tmp2->val + ret->val > 9) {
                ret->val = (tmp1->val + tmp2->val + ret->val) % 10;
                ret->next = new ListNode(1, nullptr);
                ret = ret->next;
            }
        }

        return &head;
    }
};